File: Solutions/0946/0946.py
# -*- coding:utf-8 -*-
# Author: washing
# DateTime: 2022/08/31 11:11
# File: 0946.py
# Desc: 


class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        temp = []
        while popped:
            if temp and temp[-1] == popped[0]:
                temp.pop()
                popped.pop(0)
            elif pushed:
                temp.append(pushed.pop(0))
            else: break
        return temp == popped and len(temp) <= 1

File: Solutions/0946/0946.c
/***************************************************************************************************
 * Copyright: washing
 * FileName: 0946.c
 * Author: washing
 * Version: 1.0
 * Date: 2022-8-31 11:52:57
***************************************************************************************************/


bool validateStackSequences(int* pushed, int pushedSize, int* popped, int poppedSize){
    int * temp = (int *)malloc(sizeof(int) * pushedSize);
    int pushidx=0, popidx=0, tempidx=0;
    while (popidx < poppedSize) {
        if (tempidx > 0 && temp[tempidx-1] == popped[popidx]) {
            tempidx--;
            popidx++;
        } else if (pushidx < pushedSize) {
            temp[tempidx++] = pushed[pushidx++];
        } else break;
    }
    return tempidx == 0 || (tempidx==1 && temp[0] == popped[popidx-1]);
}
